storage/redis_adapter.go

```diff
@@ -24,35 +42,64 @@ func NewRedisStorageAdapter(bufferSize int) (*redisAdapter, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &redisAdapter{
+	rsa := &redisAdapter{
 		bufferSize: bufferSize,
 		redisClient: r.NewClient(&r.Options{
 			Addr:     fmt.Sprintf("%s:%d", cfg.RedisHost, cfg.RedisPort),
 			Password: cfg.RedisPassword, // "" == no password
 			DB:       int64(cfg.RedisDB),
 		}),
-	}, nil
+		messageChannel: make(chan *message),
+	}
+	go rsa.start()
+	return rsa, nil
 }
 
-// Write adds a log message to to an app-specific list in redis using ring-buffer-like semantics
-func (a *redisAdapter) Write(app string, message string) error {
-	// Note: Deliberately NOT using MULTI / transactions here since in this implementation of the
-	// redis client, MULTI is not safe for concurrent use by multiple goroutines. It's been advised
-	// by the authors of the gopkg.in/redis.v3 package to just use pipelining when possible...
-	// and here that is technically possible. In the WORST case scenario, not having transactions
-	// means we may momentarily have more than the desired number of log entries in the list /
-	// buffer, but an LTRIM will eventually correct that, bringing the list / buffer back down to
-	// its desired max size.
-	pipeline := a.redisClient.Pipeline()
-	if err := pipeline.RPush(app, message).Err(); err != nil {
-		return err
-	}
-	if err := pipeline.LTrim(app, int64(-1*a.bufferSize), -1).Err(); err != nil {
-		return err
+func (a *redisAdapter) start() {
+	a.resetMessageQueue()
+	for {
+		select {
+		case message := <-a.messageChannel:
+			if err := a.pipeline.RPush(message.app, message.messageBody).Err(); err == nil {
+				a.queuedApps[message.app] = true
+				a.messageCount++
+				if a.messageCount == 50 {
+					a.execPipeline()
+				}
+			} else {
+				log.Printf("Error adding rpush to %s to the pipeline: %s", message.app, err)
+			}
+		case <-a.timeoutTicker.C:
+			a.execPipeline()
+		}
 	}
-	if _, err := pipeline.Exec(); err != nil {
-		return err
+}
+
+func (a *redisAdapter) execPipeline() {
+	for app := range a.queuedApps {
+		if err := a.pipeline.LTrim(app, int64(-1*a.bufferSize), -1).Err(); err != nil {
+			log.Printf("Error adding ltrim of %s to the pipeline: %s", app, err)
+		}
 	}
+	pipeline := a.pipeline
+	go func() {
+		if _, err := pipeline.Exec(); err != nil {
+			log.Printf("Error executing pipeline: %s", err)
+		}
+	}()
+	a.resetMessageQueue()
```

Comment:
Calling attention to the above: this is intended to take advantage of whatever connection pooling optimizations (if any) can be made. It executes the pipeline asynchronously while getting the ball rolling on a new pipeline.